### PR TITLE
Simplify RpcWorkerChannel Initialization in PlaceholderMode

### LINKED ISSUE #7791: Introduce list of language workers to start in placeholder mode
Add new environment variable  that lists langauge workers instead of hardcoding the list here: https://github.com/Azure/azure-functions-host/blob/443869e7a41c820288d0d5cb87b31e7569ed118d/src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs#L26-L40

This is useful to set explicit list of languages in mesh images used for Linux Consumption  where we do bring up python and node language workers.

### PR DESCRIPTION
**Pull request by pragnagopa**
### Issue describing the changes in this PR

resolves #7791

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

- Remove legacy code and OS restrictions
- Simplify which language workers host should start in placeholder mode by introducing new Environment Setting: `FUNCTIONS_WORKER_RUNTIME_PLACEHOLDERMODE_LIST` 
- If `FUNCTIONS_WORKER_RUNTIME` and `FUNCTIONS_WORKER_RUNTIME_PLACEHOLDERMODE_LIST` are set , host will start all the language workers mentioned in `FUNCTIONS_WORKER_RUNTIME` and `FUNCTIONS_WORKER_RUNTIME_PLACEHOLDERMODE_LIST`
- Related changes in Linux Consumption Images : https://github.com/Azure/azure-functions-docker-private/pull/241

### CODE REVIEW

File: src/WebJobs.Script/Environment/EnvironmentExtensions.cs [hunk 1]
```diff
@@ -481,5 +482,18 @@ public static bool IsWorkerDynamicConcurrencyEnabled(this IEnvironment environme
             }
             return false;
         }
+
+        public static HashSet<string> GetLanguageWorkerListToStartInPlaceholder(this IEnvironment environment)
+        {
+            string placeholderList = environment.GetEnvironmentVariableOrDefault(RpcWorkerConstants.FunctionWorkerPlaceholderModeListSettingName, string.Empty);
+            var placeholderRuntimeSet = new HashSet<string>(placeholderList.Split(';'));
```
Comment (ahmelsayed) [inline src/WebJobs.Script/Environment/EnvironmentExtensions.cs]: nit: 
```suggestion
            var placeholderRuntimeSet = new HashSet<string>(placeholderList.Split(';', StringSplitOptions.RemoveEmptyEntries).Select(p => p.Trim());
```

or `.Split(';', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)` in .NET 5.

File: src/WebJobs.Script/Environment/EnvironmentExtensions.cs [hunk 2]
```diff
@@ -481,5 +482,18 @@ public static bool IsWorkerDynamicConcurrencyEnabled(this IEnvironment environme
             }
             return false;
         }
+
+        public static HashSet<string> GetLanguageWorkerListToStartInPlaceholder(this IEnvironment environment)
+        {
+            string placeholderList = environment.GetEnvironmentVariableOrDefault(RpcWorkerConstants.FunctionWorkerPlaceholderModeListSettingName, string.Empty);
```
Comment (pgombar) [inline src/WebJobs.Script/Environment/EnvironmentExtensions.cs:496]: Question: where do these environment variables get populated from in Windows? For example, for Linux it's a matter of adding them in the docker image, but for Windows?

Comment (pragnagopa) [reply src/WebJobs.Script/Environment/EnvironmentExtensions.cs:496]: On Windows - we want to restrict starting a single language worker process - as all the runtimes are available on the windows VMs.  This environment setting will not be set in Windows.

**Review by pgombar — DISMISSED** [review]

File: src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs [hunk 3]
```diff
@@ -127,79 +98,22 @@ internal async Task InitializeRpcServerAsync()
...
-
-            return Task.CompletedTask;
-        }
-
-        internal bool ShouldStartStandbyPlaceholderChannels()
-        {
-            if (string.IsNullOrEmpty(_workerRuntime) && _environment.IsPlaceholderModeEnabled())
-            {
-                if (_environment.IsLinuxAzureManagedHosting())
+                if (_placeholderLanguageWorkersList != null && _placeholderLanguageWorkersList.Count() != 0)
```
Comment (balag0) [inline src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs]: I think we can remove the null check for _placeholderLanguageWorkersList. also lets log the list of runtimes to start in placeholder mode even if list is empty (for alerting).

**Comment by balag0** [comment]
small note on the timing of changes. the dockerfile changes need to go first before merging this change.

File: src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs (hunk 3 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs]: done!

**Comment by pragnagopa** [comment]
> small note on the timing of changes. the dockerfile changes need to go first before merging this change.

Yes, Docker Files PR that add new environment variable: https://github.com/Azure/azure-functions-docker-private/pull/241 --> do take a look. Will merge once approved.

File: src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs [hunk 4]
```diff
@@ -127,79 +98,28 @@ internal async Task InitializeRpcServerAsync()
...
-        }
-
-        private Task InitializePlaceholderChannelsAsync()
-        {
-            if (_environment.IsLinuxAzureManagedHosting())
+            if (_environment.IsPlaceholderModeEnabled())
             {
-                return InitializePlaceholderChannelsAsync(OSPlatform.Linux);
-            }
+                _logger.LogDebug("Initializing language worker channels '{placholderChannelList}' in placeholder mode.", string.Join(",", _placeholderLanguageWorkersList));
```
Comment (pgombar) [inline src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs]: typo `placholderChannelList` -> `placeholderChannelList`

Comment (balag0) [reply src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs]: if we can have the worker_runtime logged in this line would make writing queries easier. alert can validate
if runtime=dotnet=>  dotnet,python,node are started
node => only node is started , ..

Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs]: done!

File: src/WebJobs.Script/Workers/Rpc/RpcWorkerConstants.cs [hunk 5]
```diff
@@ -8,6 +8,8 @@ namespace Microsoft.Azure.WebJobs.Script.Workers.Rpc
     public static class RpcWorkerConstants
     {
         public const string FunctionWorkerRuntimeSettingName = "FUNCTIONS_WORKER_RUNTIME";
+        // semicolon seperated string - list of runtimes to start in placeholder mode
+        public const string FunctionWorkerPlaceholderModeListSettingName = "FUNCTIONS_WORKER_RUNTIME_PLACEHOLDERMODE_LIST";
```
Comment (brettsam) [inline src/WebJobs.Script/Workers/Rpc/RpcWorkerConstants.cs]: Is this being set in all of our environments now?

Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/RpcWorkerConstants.cs]: Please see https://github.com/Azure/azure-functions-docker-private/pull/241 

Starting multiple language workers in placeholder mode only applies to Linux Consumption - as we need to optimize existing pool due to capacity restrictions. On Windows - only a single language worker channel is started in placeholder mode - this is controlled via `FUNCTIONS_WORKER_RUNTIME` that is set on the placeholder template sites in Antares.

Do let me know if you have any other concerns.

**Comment by pragnagopa** [comment]
@brettsam  / @balag0  - gentle ping for approval. Addressed CR

**Comment by balag0** [comment]
> @brettsam / @balag0 - gentle ping for approval. Addressed CR

Looks like we need to retarget to dev branch? (pr shows > 700 commits now)

**Comment by pragnagopa** [comment]
> Looks like we need to retarget to dev branch? (pr shows > 700 commits now)

Thanks! Rebased and fixed target branch !

**Review by balag0 — DISMISSED** [review]

File: src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs [hunk 6]
```diff
@@ -127,79 +98,28 @@ internal async Task InitializeRpcServerAsync()
             }
             catch (Exception grpcInitEx)
             {
-                var hostInitEx = new HostInitializationException($"Failed to start Rpc Server. Check if your app is hitting connection limits.", grpcInitEx);
+                throw new HostInitializationException($"Failed to start Rpc Server. Check if your app is hitting connection limits.", grpcInitEx);
```
Comment (brettsam) [inline src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs:101]: Interesting... were we ignoring this exception previously?

File: src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs [hunk 7]
```diff
@@ -127,79 +98,28 @@ internal async Task InitializeRpcServerAsync()
...
+                throw new ArgumentNullException(nameof(_placeholderLanguageWorkersList));
             }
 
-            return InitializeWebHostRuntimeChannelsAsync();
-        }
-
-        private Task InitializePlaceholderChannelsAsync()
-        {
-            if (_environment.IsLinuxAzureManagedHosting())
+            if (_environment.IsPlaceholderModeEnabled())
```
Comment (brettsam) [inline src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs:112]: So is this whole method now a no-op unless in placeholder mode? That's now how it used to work, is it? I see it used to eventually call into `InitializeWebHostRuntimeChannelsAsync()`

File: test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/SamplesEndToEndTests_Java.cs [hunk 8]
```diff
@@ -50,8 +50,8 @@ public async Task JavaProcess_Same_AfterHostRestart()
             Assert.True(javaProcessesAfter.Count() > 0);
             // Verify number of java processes before and after restart are the same.
             Assert.Equal(javaProcessesBefore.Count(), javaProcessesAfter.Count());
-            // Verify Java same java process is used after host restart
-            var result = javaProcessesBefore.Where(pId1 => !javaProcessesAfter.Any(pId2 => pId2 == pId1));
+            // Verify Java different java process is used after host restart
+            var result = javaProcessesBefore.Where(pId1 => javaProcessesAfter.Any(pId2 => pId2 == pId1));
```
Comment (brettsam) [inline test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/SamplesEndToEndTests_Java.cs:54]: Is this expected? It's not clear how this change is affecting this scenario. (Also, should change the test name if it's expected).

File: src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs (hunk 6 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs:101]: I assumed this was an oversight and fixed it !

File: src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs (hunk 7 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs:112]: Correct ! We only want to start Language Worker Channels in "WebHost" Scope only when running in Azure - Placeholder Mode

File: test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/SamplesEndToEndTests_Java.cs (hunk 8 above)
Comment (pragnagopa) [reply test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/SamplesEndToEndTests_Java.cs:54]: Yes, this expected with the change now - related to the comment above - We only want to start Language Worker Channels in "WebHost" Scope only when running in Azure - Placeholder Mode - Earlier code was unnecessarily complex with conditions per OS and also when running locally vs running in Azure

Comment (brettsam) [reply test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/SamplesEndToEndTests_Java.cs:54]: Okay -- let's change this test name then so it matches the expectation.

File: src/WebJobs.Script/Workers/Rpc/RpcWorkerConstants.cs (hunk 5 above)
Comment (liliankasem) [reply src/WebJobs.Script/Workers/Rpc/RpcWorkerConstants.cs]: nit: can we seperate `PLACEHOLDERMODE` into `PLACEHOLDER_MODE` - feels and looks cleaner if keep words seperated :)

File: test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/SamplesEndToEndTests_Java.cs (hunk 8 above)
Comment (pragnagopa) [reply test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/SamplesEndToEndTests_Java.cs:54]: Thanks! I missed method was explicit!

**Review by liliankasem — DISMISSED** [review]

File: src/WebJobs.Script/Workers/Rpc/RpcWorkerConstants.cs (hunk 5 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/RpcWorkerConstants.cs]: done! Thanks!

File: src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs (hunk 6 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs:101]: @brettsam  - OK to resolve this comment?

**Review by brettsam — APPROVED** [review]

**Pull request merged**